paleofetch.c

```diff
@@ -232,37 +232,127 @@ static char *get_uptime() {
+      return strcmp(battery_type, "Battery") == 0;
+    }
+    else {
+      return 0;
+    }
+  }
+}
+
+static struct dirent *get_nth_battery(int n) {
+  static struct dirent **power_supply_directory = NULL;
+  static int num_dirs = -1;
+
+  if (!power_supply_directory) {
+    num_dirs = scandir("/sys/class/power_supply/", &power_supply_directory, directory_is_battery, alphasort);
+  }
+
+  if (n < num_dirs)
+    return power_supply_directory[n];
+  else
+    return NULL;
+}
+
 // returns "<Battery Percentage>% [<Charging | Discharging | Unknown>]"
 // Credit: allisio - https://gist.github.com/allisio/1e850b93c81150124c2634716fbc4815
-static char *get_battery_percentage() {
-  int battery_capacity;
-  FILE *capacity_file, *status_file;
-  char battery_status[12] = "Unknown";
+// returns percentage of the nth battery
+// changes the contents of *label to the name of the battery
+static char *get_nth_battery_percentage(int n, char** label) {
+  struct dirent *battery_directory = get_nth_battery(n);
 
-  if ((capacity_file = fopen(BATTERY_DIRECTORY "/capacity", "r")) == NULL) {
-    status = ENOENT;
-    halt_and_catch_fire("Unable to get battery information");
+  if (battery_directory == NULL)
+    return calloc(1, 1);
+
+  // could not figure out possible values for this
+  char battery_capacity[12];
+
+  FILE *capacity_file, *status_file, *model_file;
+  char battery_status[12] = "Unknown";
+  
+  char file_path_buffer[BUF_SIZE];
+  snprintf(file_path_buffer, sizeof(file_path_buffer), "/sys/class/power_supply/%s/", battery_directory->d_name);
+  
+  char *file_path_buffer_end = file_path_buffer + strlen(file_path_buffer);
+
+  bool capacity_is_percent = true;
+
+  strcpy(file_path_buffer_end, "capacity");
+  if ((capacity_file = fopen(file_path_buffer, "r")) == NULL) {
+    strcpy(file_path_buffer_end, "capacity_level");
+    if ((capacity_file = fopen(file_path_buffer, "r")) == NULL) {
+      status = ENOENT;
+      halt_and_catch_fire("Unable to get battery information");
+    }
+    capacity_is_percent = false;
   }
 
-  fscanf(capacity_file, "%d", &battery_capacity);
+  fscanf(capacity_file, "%s", battery_capacity);
   fclose(capacity_file);
 
-  if ((status_file = fopen(BATTERY_DIRECTORY "/status", "r")) != NULL) {
+  strcpy(file_path_buffer_end, "status");
+  if ((status_file = fopen(file_path_buffer, "r")) != NULL) {
     fscanf(status_file, "%s", battery_status);
     fclose(status_file);
   }
 
   // max length of resulting string is 19
   // one byte for padding incase there is a newline
   // 100% [Discharging]
+  // High [Discharging]
   // 1234567890123456789
   char *battery = malloc(20);
 
-  snprintf(battery, 20, "%d%% [%s]", battery_capacity, battery_status);
+  if (capacity_is_percent) {
+    snprintf(battery, 20, "%s%% [%s]", battery_capacity, battery_status);
+  }
+  else {
+    snprintf(battery, 20, "%s [%s]", battery_capacity, battery_status);
+  }
+
+  // get model name of battery
+  char *model_name = malloc(30);
```

Comment:
Where is this allocated memory freed?

Reply:
Hmm, I hadn't thought about this. Since not every field's label is dynamically allocated, the label's memory cannot be freed indiscriminately after output like we do with the values. Do you have any suggestions on how we could free the memory cleanly, without just adding special cases for all of the functions that _do_ dynamically allocate memory?